Mathlib/Combinatorics/SimpleGraph/VertexCover.lean

```diff
@@ -0,0 +1,188 @@
+
+@[simp]
+theorem vertexCoverNum_of_subsingleton [Subsingleton V] : vertexCoverNum G = 0 := by
+  simp [SimpleGraph.subsingleton_iff.mpr _ |>.allEq G ⊥]
+
+@[simp]
+theorem vertexCoverNum_eq_zero : vertexCoverNum G = 0 ↔ G = ⊥ := by
+  constructor
+  · intro h
+    have := vertexCoverNum_exists G
+    simp_all
+  · simp_all
+
+theorem vertexCoverNum_le_card_sub_one : vertexCoverNum G ≤ ENat.card V - 1 := by
+  nontriviality V
+  obtain ⟨x⟩ := not_subsingleton_iff_nontrivial.mp (not_subsingleton V) |>.to_nonempty
+  refine ENat.forall_natCast_le_iff_le.mp fun n hn ↦ ?_
+  simp only [vertexCoverNum, le_iInf_iff, Subtype.forall] at hn
+  have := hn (Set.univ \ {x}) (by grind [IsVertexCover, Adj.ne'])
+  simpa [Set.encard_diff_singleton_of_mem (Set.mem_univ _)] using this
+
+@[simp]
+theorem vertexCoverNum_ne_top_of_finite [Finite V] : vertexCoverNum G ≠ ⊤ :=
+  ne_top_of_le_ne_top (by simpa) (@vertexCoverNum_le_card_sub_one V G)
+
+theorem vertexCoverNum_le_encard_edgeSet : vertexCoverNum G ≤ G.edgeSet.encard := by
+  by_cases h' : G.edgeSet = ∅
+  · simp [h', SimpleGraph.edgeSet_eq_empty.mp]
+  have := ne_bot_iff_exists_adj.mp (by simpa using h')
+  refine ENat.forall_natCast_le_iff_le.mp fun n hn ↦ ?_
+  simp only [vertexCoverNum, le_iInf_iff, Subtype.forall] at hn
+  have := hn ((·.out.1) '' G.edgeSet) (fun v w hadj ↦ by
+    have := Sym2.mem_iff.mp <| Sym2.out_fst_mem s(v, w)
+    grind [mem_edgeSet])
+  grw [this]
+  exact Set.encard_image_le (fun x ↦ (Quot.out x).1) G.edgeSet
+
+@[simp]
+theorem vertexCoverNum_ne_top_of_edgeSet_finite (h : G.edgeSet.Finite) : vertexCoverNum G ≠ ⊤ :=
+  ne_top_of_le_ne_top (Set.encard_ne_top_iff.mpr h) (@vertexCoverNum_le_encard_edgeSet V G)
+
+theorem vertexCoverNum_top : vertexCoverNum (completeGraph V) = ENat.card V - 1 := by
+  by_cases h' : Subsingleton V
+  · simp [tsub_eq_zero_of_le]
+  rw [not_subsingleton_iff_nontrivial] at h'
+  obtain ⟨x⟩ := h'.to_nonempty
+  refine ENat.eq_of_forall_natCast_le_iff fun n ↦ ⟨fun hn ↦ ?_, fun hn ↦ ?_⟩
+  · grw [hn, vertexCoverNum_le_card_sub_one]
+  · by_contra! hh
+    have : n - 1 ≤ ENat.card V := by
+      grw [tsub_le_iff_right, hn]
+      simp [add_assoc, one_add_one_eq_two]
+    obtain ⟨t, ht₁, ht₂⟩ := exists_of_le_vertexCoverNum (n - 1) (ENat.le_sub_one_of_lt hh) this
+    have : 1 < (Set.univ \ t).encard := by
+      refine ENat.add_one_le_iff (by simp) |>.mp ?_
+      rw [Set.encard_diff (by simp) (Set.finite_of_encard_eq_coe ht₁), Set.encard_univ]
+      refine ENat.le_sub_of_add_le_left (by simp [ht₁]) ?_
+      refine add_le_of_le_tsub_right_of_le (Order.add_one_le_of_lt ENat.one_lt_card) ?_
+      grw [ht₁, ENat.coe_sub, hn]
+      simp [add_assoc, one_add_one_eq_two, le_tsub_add]
+    obtain ⟨a, b, _, _, hne⟩ := Set.one_lt_encard_iff.mp <| this
+    have := @ht₂ a b (by simp [hne])
+    grind
+
+theorem vertexCoverNum_le_of_relHom (f : G →g H) (hf : Function.Injective f) :
+    vertexCoverNum G ≤ vertexCoverNum H := by
```

Comment:
Can you deduce monotonicity of `vertexCoverNum` (for graphs on the same vertices) and tag it with gcongr?